### PR TITLE
Resolve GPDB_12_MERGE_FIXMEs in motion/tupser.c

### PR DESCRIPTION
**Pull request by interma**
There are two fixmes in this PR:
### Removed `s_tupSerMemCtxt`
This memory context is used in `SerializeRecordCacheIntoChunks()`, and its prerequisite is
https://github.com/greenplum-db/gpdb/blob/b3acafa5cbcdc782461553da6f24b2bf3342bdbf/src/backend/cdb/motion/cdbmotion.c#L1249-L1255

So it's only for the **Transient RECORD** types (an example: https://github.com/greenplum-db/gpdb/issues/4444), and means not every query contains it and memory consumption is not big. 

After removed it, the related memory is alloced in MotionLayerMemCtxt (and parent is ExecutorState): get freed when the query is finished. So it's safe to remove `s_tupSerMemCtxt`.

### MinimalTuple vs. MemTuple in motion
Postgres introduced MinimalTuple, and GPDB introduced MemTuple. Their usage scenarios are similar. In 6x, we used MemTuple (gpdb introduced it) for the data transportation of motion. 

In the process of gp7's merge work, most of MemTuple scenarios are replaced with MinimalTuple (include motion).
But seems using MemTuple has a little performance boost, pls see the discussion:
https://app.slack.com/client/T6T3AS6NL/C74P7FZFG/thread/C74P7FZFG-1582306400.252200

We need to be cautious here, so don't change it in this PR.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

### CODE REVIEW

**Comment by interma** [comment]
ready for review now.

**Comment by interma** [comment]
A concern of `s_tupSerMemCtxt`: if the `NextRecordTypmod` variable is proportional to the tuple number of a query, if so, it will consume many memories.
I will verify it later, thanks zhenghua for mentioning it.

File: src/backend/cdb/motion/tupser.c [hunk 1]
```diff
@@ -405,6 +381,9 @@ SerializeTuple(TupleTableSlot *slot, SerTupInfo *pSerInfo, struct directTranspor
 	 * GPDB_12_MERGE_FIXME: This used to support serializing memtuples directly.
 	 * That got removed with MinimalTuples in the merge. Resurrect the MemtUple
 	 * support if there's a performance benefit.
+	 *
+	 * Details please see slack discussion:
+	 * https://app.slack.com/client/T6T3AS6NL/C74P7FZFG/thread/C74P7FZFG-1582306400.252200
```
Comment (adam8157) [inline src/backend/cdb/motion/tupser.c:386]: slack is not public, someone who wants to read the details still needs to register, maybe have some brief summary directly?

Comment (interma) [reply src/backend/cdb/motion/tupser.c:386]: got it, the discussion is a lit long, I will try to summarize it.

**Comment by interma** [comment]
> A concern of `s_tupSerMemCtxt`: if the `NextRecordTypmod` variable is proportional to the tuple number of a query, if so, it will consume many memories.
> I will verify it later, thanks zhenghua for mentioning it.

I checked the code, `assign_record_type_typmod()` may be called on each tuple, but it only registers a new type when not find the type in `RecordCacheHash`: 
https://github.com/greenplum-db/gpdb/blob/c394f5650bee38c70a7e11ede97d96b60873e165/src/backend/utils/cache/typcache.c#L1802-L1823

So `NextRecordTypmod` doesn't increase too much in one query.

File: src/backend/cdb/motion/tupser.c (hunk 1 above)
Comment (interma) [reply src/backend/cdb/motion/tupser.c:387]: done

File: src/backend/cdb/motion/tupser.c [hunk 2]
```diff
@@ -403,8 +379,12 @@ SerializeTuple(TupleTableSlot *slot, SerTupInfo *pSerInfo, struct directTranspor
 
 	/*
 	 * GPDB_12_MERGE_FIXME: This used to support serializing memtuples directly.
```
Comment (adam8157) [inline src/backend/cdb/motion/tupser.c:381]: Greenplum uses minimal tuples now, and will not change it back.

I don't see a reason to keep this FIXME, explaining things in the commit message is good enough to me.

**Review by adam8157 — APPROVED** [review]
IMO the other FIXME could be removed too.

**Comment by soumyadeep2007** [comment]
@interma can we merge this or do we need more eyes on this?

**Comment by kainwen** [comment]
> @interma can we merge this or do we need more eyes on this?

I have a concern told to hin not resolve yet. "Will other session keep change type record info and lead to oom？ Since we do not reset this part of memory per tuple as before."

**Comment by interma** [comment]
Yes, I believe there is no `s_tupSerMemCtxt` in "ancient" time of gp, but it consumes too much memory in some scenarios , then one "antecessor" introduce it. But I haven't find out the scenario yet.

(And I checked it in gp-historical repo, it's existed in the original version of gp4, so cannot find the commit message)

**Comment by interma** [comment]
After rechecking, still not 100% sure it's safe to remove `s_tupSerMemCtxt`.
I think we should prefer safe over a clear code here, so don't remove it and close this PR.

**Pull request closed without merging**